extra/mariabackup/xtrabackup.cc

```diff
@@ -4727,18 +4768,10 @@ bool Backup_datasinks::backup_low()
 		} else {
 			msg("Error: recv_sys.find_checkpoint() failed.");
 		}
-
 		recv_sys.lsn = lsn;
-		stop_backup_threads();
-	}
-
-	if (metadata_to_lsn && xtrabackup_copy_logfile()) {
-		mysql_mutex_unlock(&recv_sys.mutex);
-		ds_close(dst_log_file);
-		dst_log_file = NULL;
-		return false;
 	}
 
+	stop_backup_threads();
```

Comment:
Hmm.. I don't think I agree.

The important distinction the existing code makes is that the background is ensured to copy enough redo log (metadata_to_lsn) so that the backup is consistent. After this point we are free to decide how much more log we want to copy and the current behaviour is to copy as much. 

There are different view points on deciding how much is good to copy here (usually replication provisioning wants it as close as the master) but the current code makes it easy to manoeuvre in case we want to do something else. Thus I don't see a point changing it.

Reply:
The variable name `metadata_to_lsn` was confusing me. The name is somewhat logical there when referring to `--prepare --incremental`, where we’d first have to apply `.delta` files to get from the previous backed-up LSN (`metadata_from_lsn`) to the checkpoint LSN of the current `ib_logfile0` (`metadata_to_lsn`).

Initially, `xtrabackup_backup_func()` will read the *initial* checkpoint LSN into `log_sys.next_checkpoint_lsn` and write it to the backup `ib_logfile0` header. Before this change, the `metadata_to_lsn` would be assigned to the *final* checkpoint LSN at the end of backup. The only call of this function `Backup_datasinks::backup_low()` is from `BackupStages::stage_block_commit()`, which is the most escalated locking state before `stage_end()`. 

There are two requirements here. We must copy the log past the current LSN of the server, and we must ensure that no data page that we copied carries a newer `FIL_PAGE_LSN` than that.

A scenario to consider would be along the following lines:
1. `backup_wait_for_lsn()` is invoked on `get_current_lsn(mysql_connection)`, which would parse the `Log sequence number` section of the output of `SHOW ENGINE INNODB STATUS` from the server. With dafdb55c927403db40adc5f247bb4fe2390505a5 this LSN would be determined already in `BackupStages::stage_block_ddl()` and not `BackupStages::stage_block_commit()`. You seem to be right, the purpose of `BACKUP STAGE BLOCK_COMMIT` seems to be to get a consistent state of all storage engines, and indeed we should determine the target LSN after acquiring that lock.
2. Could some data pages with an even larger LSN have been copied? It seems to me that the copying (`InnodbDataCopier innodb_data_copier`) would already have finished at `BackupStages::stage_start()`.
3. The server advances the checkpoint LSN, and `Backup_datasinks::backup_low()` finds it in the server’s `ib_logfile0` by invoking `recv_sys.find_checkpoint()`. By copying the log at least until this LSN, we’d ensure that no data page that we copied contains a ‘too new’ `FIL_PAGE_LSN`.

The last point seems to be an overkill; it seems that we should set the backup target LSN to `get_current_lsn(mysql_connection)` and adjust it further to not be less than any `FIL_PAGE_LSN` that we had copied from the data files. There should be no need to re-read the checkpoint headers of the server’s `ib_logfile0`.

Reply:
According to the design in [MDEV-5336](https://jira.mariadb.org/browse/MDEV-5336) there is not that much difference between `BLOCK_DDL` and `BLOCK_COMMIT` from the InnoDB point of view. The design also clearly says that the logical time of the backup is after `BACKUP STAGE BLOCK_COMMIT` has been successfully executed. If we decided to ‘freeze’ InnoDB already between `BLOCK_DDL` and `BLOCK_COMMIT`, then we’d face inconsistencies with regard to the following:
* The binary log
* Non-InnoDB system tables
* `ENGINE=RocksDB` tables (`rocksdb_create_checkpoint()` is executed inside `BackupStages::stage_block_commit()`)

Reply:
I revised the solution. The main difference is that by the time `Backup_datasinks::backup_low()` is executed, the `log_copying_thread` will already have stopped, and it will have waited for `metadata_last_lsn`. This could result in slightly less context switching and some less copying of the log. The logic related to `metadata_to_lsn` (waiting for the latest checkpoint LSN) was preserved.